### PR TITLE
Add negation to selector syntax.

### PR DESCRIPTION
**Pull request by fasaxc**
Added the `!` operator, which negates a sub-expression.

### CODE REVIEW

File: calico/felix/test/test_selectors.py [hunk 1]
```diff
@@ -61,23 +61,34 @@ def test_matches():
     yield check_no_match, 'a not in {"a"}', {"a": "a"}
     yield check_no_match, 'a in {"a", "b"}', {"a": "c"}
     yield check_no_match, 'has(b)', {"a": "b"}
+    yield check_no_match, '!!has(b)', {"a": "b"}
+    yield check_no_match, '! has(a)', {"a": "b"}
+    yield check_no_match, '!has(a)', {"a": "b"}
+    yield check_no_match, '!!! has(a)', {"a": "b"}
+    yield check_no_match, '!!!has(a)', {"a": "b"}
+    yield check_no_match, '!! ! has(a)', {"a": "b"}
+    yield check_no_match, '! !!has(a)', {"a": "b"}
```
Comment (matthewdupre) [inline calico/felix/test/test_selectors.py]: We should probably check that it's possible to get a positive match when ! is used.

**Comment by matthewdupre** [comment]
Code here looks fine.  The !! simplifier seems a bit superfluous, but it's simple enough.

According to the datamodel we also support ! for specifying rule targets (i.e. src_selector).  I might have missed something, but I don't think that code can deal with complex nesting of !s (e.g. it will need to make the translation !(X && Y) -> !X || !Y) for programming into iptables?  Am I missing something, or do we still need to write some code to support this?

**Comment by matthewdupre** [comment]
Also - thanks for writing this so quickly! :)

**Comment by fasaxc** [comment]
@matthewdupre we actually render the whole selector into an ipset rather than rendering its individual matches as separate sets.  Made that choice precisely for the reason you outlined, it's tricky to do all the boolean ops in iptables.

**Comment by fasaxc** [comment]
retest this please

**Comment by fasaxc** [comment]
@matthewdupre FTR, the '!!' simplifier isn't strictly necessary but I think it's a lot more efficient to do operators as a `ZeroOrMore(...)` list rather than using recursion in pyparsing so you end up sending the 0-case and the 1-case through the same parse action anyway; may as well handle the n-case too.

**Pull request merged**